File: 이상훈/1주차/2회차/홀짝_연결리스트.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode *odd = head;
        ListNode *even = head == nullptr? nullptr : head->next;
        ListNode *evenFirst = even;

        if (odd == nullptr || even == nullptr || even->next == nullptr) return head;


        while (odd->next != nullptr  && even->next != nullptr ) {

            if (odd->next == even) {
                odd->next = even->next;
                odd = even->next;
            } else {
                even->next = odd->next;
                even = odd->next;
            }
        };

        even->next = nullptr;
        odd->next = evenFirst;


        return head;
    }
};

File: 이상훈/1주차/2회차/홀짝_연결리스트.py
# https://leetcode.com/problems/odd-even-linked-list
# https://sanghoonly.tistory.com/16?category=990444


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 본 코드는 odd 노드와 even 노드를 구분하지 않고 구현한 코드.
def oddEvenList(head: Optional[ListNode]) -> Optional[ListNode]:
    if not head or not head.next:
        return head

    even_head, cur, prev = head.next, head, None
    length = 1

    while cur and cur.next:
        prev = cur
        next = cur.next
        cur.next = cur.next.next
        cur = next
        length += 1

    if length % 2:
        cur.next = even_head
    else:
        prev.next = even_head

    return head
